adr/ADR-43.md

```diff
@@ -0,0 +1,87 @@
+# JetStream Per-Message TTL
+
+| Metadata | Value                     |
+|----------|---------------------------|
+| Date     | 2024-07-11                |
+| Author   | @ripienaar                |
+| Status   | Approved                  |
+| Tags     | jetstream, client, server |
+
+## Context and motivation
+
+Streams support a one-size-fits-all approach to message TTL based on the MaxAge setting. This causes any message in the 
+Stream to expire at that age.
+
+There are numerous uses for a per-message version of this limit, some listed below:
+
+ * KV tombstones are a problem in that they forever clog up the buckets with noise, these could have a TTL to make them expire once not useful anymore
+ * Server-applied limits can result in tombstones with a short per message TTL so that consumers can be notified of limits being processed. Useful in KV watch scenarios being notified about TTL removals
+ * A stream may have a general MaxAge but some messages may have infinite retention, think a schema or type hints in a KV bucket that is forever while general keys have TTLs
+ * TODO: user use cases
+
+Related issues [#3268](https://github.com/nats-io/nats-server/issues/3268)
+
+## Per-Message TTL
+
+We will allow a message to supply a TTL as a duration in a header called `Nats-TTL` followed by the duration as 
+nanoseconds, the sever may not support nanosecond level precision and clients can choose to expose TTL as
+seconds or nanoseconds.
+
+The duration will be used by the server to calculate the deadline for removing the message based on its Stream 
+timestamp and the stated duration.
+
+The TTL may not exceed the Stream MaxAge to give administrators a hard cap for the purposes of managing resources.
```

Comment:
Actually - you could from the description earlier - it would be useful to have a TTL that is beyond the max age, that way you can have a permalink message, whether others are pruned as the lifecycle continues. Sounds non-intuitive, but allows default (no headers set) to clean the messages as they age, but the explicit ones remain.